include/circt/Dialect/SMT/SMTOps.td

```diff
@@ -16,8 +16,152 @@ include "mlir/IR/EnumAttr.td"
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>
+]> {
+  let summary = "declare a symbolic value of a given sort";
+  let description = [{
+    This operation declares a symbolic value just as the `declare-const`
+    statement in SMT-LIB 2.6. The result type determines the SMT sort of the
+    symbolic value. The returned value can then be used to refer to the symbolic
+    value instead of using the identifier like in SMT-LIB.
+
+    The optionally provided string will be used as a prefix for the newly
+    generated identifier (useful for easier readability when exporting to
+    SMT-LIB). Each `declare_const` will always provide a unique new symbolic
+    value even if the identifier strings are the same.
+
+    Note that this operation cannot be marked as Pure since two operations (even
+    with the same identifier string) could then be CSEd, leading to incorrect
+    behavior.
+  }];
+
+  let arguments = (ins OptionalAttr<StrAttr>:$namePrefix);
+  let results = (outs Res<AnySMTType, "a symbolic value", [MemAlloc]>:$result);
+
+  let assemblyFormat = [{
+    ($namePrefix^)? attr-dict `:` qualified(type($result))
+  }];
+}
+
+def SolverOp : SMTOp<"solver", [
+  IsolatedFromAbove,
+  SingleBlockImplicitTerminator<"smt::YieldOp">,
+]> {
+  let summary = "create a solver instance within a lifespan";
+  let description = [{
+    This operation defines an SMT context with a solver instance. SMT operations
+    are only valid when being executed between the start and end of the region
+    of this operation. Any invocation outside is undefined. However, they do not
+    have to be direct children of this operation. For example, it is allowed to
+    have SMT operations in a `func.func` which is only called from within this
+    region. No SMT value may enter or exit the lifespan of this region (such
+    that no value created from another SMT context can be used in this scope and
+    the solver can deallocate all state required to keep track of SMT values at
+    the end).
+
+    As a result, the region is comparable to an entire SMT-LIB script, but
+    allows for concrete operations and control-flow. Concrete values may be
+    passed in and returned to influence the computations after the `smt.solver`
+    operation.
+
+    Example:
+    ```mlir
+    %0:2 = smt.solver (%in) {smt.some_attr} : (i8) -> (i8, i32) {
+    ^bb0(%arg0: i8):
+      %c = smt.declare_const "c" : !smt.bool
+      smt.assert %c
+      %1 = smt.check sat {
+        %c1_i32 = arith.constant 1 : i32
+        smt.yield %c1_i32 : i32
+      } unknown {
+        %c0_i32 = arith.constant 0 : i32
+        smt.yield %c0_i32 : i32
+      } unsat {
+        %c-1_i32 = arith.constant -1 : i32
+        smt.yield %c-1_i32 : i32
+      } -> i32
+      smt.yield %arg0, %1 : i8, i32
+    }
+    ```
+
+    TODO: solver configuration attributes
+  }];
+
+  let arguments = (ins Variadic<AnyNonSMTType>:$inputs);
+  let regions = (region SizedRegion<1>:$bodyRegion);
```

Comment:
If you want to allow `scf.if` in your `smt.solver`, then you'll need to relax this and allow arbitrary CFGs, right?.

Reply:
Only if I want to already lower them to the CF dialect before lowering the SMT dialect, right? Not sure if we need that. On the other hand, I find it quite handy when we can just assume there is only one block.

Reply:
Oh yeah that makes sense, I also got reminded of `scf.execute_region` which can be used in case we really need that.